File: Python/113.py
from typing import List
import copy

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        result = []
        if root is None:
            return result
        resultTmp = []; Tmp = [0]
        self.func(root, sum, Tmp, result, resultTmp)

        return result
    
    def func(self, root: TreeNode, sum: int, Tmp: List[int], result: List[List[int]], resultTmp: List[int]):
        if root.left is None and root.right is None:
            Tmp[0] += root.val
            resultTmp.append(root.val)
            if Tmp[0] == sum:
                resultT = copy.deepcopy(resultTmp)
                result.append(resultT)
            return
        Tmp[0] += root.val
        resultTmp.append(root.val)
        t = Tmp[0]
        resultT = copy.deepcopy(resultTmp)
        if root.left is not None:
            self.func(root.left, sum, Tmp, result, resultTmp)
        Tmp[0] = t
        if root.right is not None:
            self.func(root.right, sum, Tmp, result, resultT)

File: C++/113.cpp
#include <iostream>
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

/*二叉树路径总和II
使用先序遍历的思路，遍历整个二叉树，其中达到叶子节点时，判断和是否等于目标
同时在遍历的过程中，维持一个列来保存经过的节点，该列在节点返回时，会吐出该
节点
 */

class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        
    }
};